### ISSUE TITLE
bin/console messenger:consume-messages amqp in PROD env using amqp_lib transport

### ISSUE THREAD

**Issue by LPodolski** [php-enqueue/messenger-adapter#23]
in documentation is stated that to consume you need to run:

```
bin/console messenger:consume-messages amqp
```

my question is how to run this in PROD environment because in doc there is not info how to create queues in rabbitMQ

running in prod like:

`messenger:consume-messages amqp --env=prod`

results in error: ` NOT_FOUND - no queue 'messages' in vhost '/' `

in dev env they are created because of debug flag:

```
if ($this->debug) {
            $this->contextManager->ensureExists($destination);
        }
```

in prod, presumably they should be crated manually

but command: `enqueue:setup-broker`

does not create queues

I found also queue is created here:

```
public function recoverException(\Exception $exception, array $destination): bool
    {
        if ($exception instanceof \AMQPQueueException) {
            if (404 === $exception->getCode()) {
                return $this->ensureExists($destination);
            }
        }

        return false;
    }
```

on exception but, my exception here is different from one that you are catching mine is:

`\PhpAmqpLib\Exception\AMQPProtocolChannelException`

from PhpAmqpLib, one of supported enqueue transports:

https://github.com/php-enqueue/enqueue-dev/tree/master/docs/transport
_______________________________________________________________

to sum it up either currently there is bug that messenger adapter is hardcoded to use only one of supported transports

or there is no info id doc how to create queues for prod environemnt

**Comment by srigi** [comment]
+1 to this. I'm also having problems running my consumers in production ENV. I get similar excp as @LPodolski `NOT_FOUND - no queue 'transactional-emails' in vhost '/'`.

There is no clear way how to pre-create queues in prod ENV. Only solution is to enable APP_DEBUG. That is just unacceptable in prod.

**Comment by salletti** [comment]
The problem is that the queue is created only when debug is active, like in dev.

File: vendor/enqueue/messenger-adapter/QueueInteropTransport.php  

```
if ($this->debug) {
      $this->contextManager->ensureExists($destination);
}
```
So maybe in prod you must use the command for create manually the queue:

https://github.com/php-enqueue/enqueue-dev/blob/master/docs/bundle/cli_commands.md#enqueuesetup-broker

**Comment by makasim** [comment]
The linked command has nothing to do with messenger 

It is part of enqueue client library

**Comment by salletti** [comment]
In my opinion the problem concerns php enqueue library.

**Comment by StarGuardian** [comment]
I bumped into the same issue and found following workaround. Wrote my own transport factory (It is basically the same except one line):
`class AmqpTransportFactory implements TransportFactoryInterface
{
    private $encoder;
    private $decoder;
    private $debug;

    public function __construct(EncoderInterface $encoder, DecoderInterface $decoder, bool $debug)
    {
        $this->encoder = $encoder;
        $this->decoder = $decoder;
        $this->debug = $debug;
    }

    public function createTransport(string $dsn, array $options): TransportInterface
    {
        return new AmqpTransport($this->encoder, $this->decoder, $this->createConnection($dsn, $options));
    }

    public function supports(string $dsn, array $options): bool
    {
        return 0 === strpos($dsn, 'amqp://');
    }

    protected function createConnection(string $dsn, array $options) {
        $connection = Connection::fromDsn($dsn, $options, $this->debug);
        $connection->setup(); // <--- Workaround is here

        return $connection;
    }
}`

Then I override messenger.transport.ampq.factory definition in my services.yml. It works well!

**Comment by tdsullivan** [comment]
I know this is an old thread, and this isn't an optimal solution, but for anyone looking for a quick workaround, adding the following to your services definition file, to default the 'debug' argument to true, seems to do the trick.
```
    enqueue.messenger_transport.factory:
        class: 'Enqueue\MessengerAdapter\QueueInteropTransportFactory'
        arguments:
            - '@messenger.default_serializer'
            - '@enqueue.locator'
            - true
        tags: ['messenger.transport_factory']
```

**Comment by tdsullivan** [comment]
Also, regarding the comments above about the official amqp ext transport, that is bundled with symfony messenger, handling this the same way, this behaviour was changed in version 4.3.

[Using AMQP auto-setup in all cases, not just in debug #30579](https://github.com/symfony/symfony/pull/30579)

**Comment by wexotht** [comment]
@sroze a fix for this would be nice. Still an issue meaning a lot of manual work on production environments upon setup
